Migrate WorksheetTemplates to DX

Bugerman58 opened the pull request:
## Description of the issue/feature this PR addresses

This PR migrates the AT WorksheetTemplate and WorksheetTemplates to Dexterity types for Python 3 compatibility.

## Current behavior before PR

WorksheetTemplate and WorksheetTemplates are content types based on the Archetypes framework

## Desired behavior after PR is merged

WorksheetTemplate and WorksheetTemplates are content types based on the Dexterity framework

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

Bugerman58 committed:
init migration for worksheet templates

Bugerman58 committed:
method for migrate

Bugerman58 committed:
fixed typo's

Bugerman58 committed:
refactoring for new methods name

Bugerman58 committed:
fix test, refactoring

Bugerman58 committed:
fixed configure adapter

Bugerman58 committed:
merge 2.x

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-tempaltes-to-dx

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-tempaltes-to-dx

Bugerman58 committed:
merge 2.x

Bugerman58 committed:
symplified upgrade step

Bugerman58 committed:
temporary fix linter warning

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-tempaltes-to-dx

Bugerman58 committed:
listing services for ws template

Bugerman58 committed:
commit local changes

Bugerman58 committed:
merge 2.x

Bugerman58 committed:
add vocabulary for analyses types

Bugerman58 committed:
local changes

Bugerman58 committed:
merge  2.x

Bugerman58 committed:
intermediate commit

Bugerman58 committed:
merge 2.x

Bugerman58 committed:
fix migrate step for upgrade

Bugerman58 committed:
fixdefault factory for template layout

Bugerman58 committed:
fix calling default factory for tempalte layout

Bugerman58 committed:
fix upgrade step for migrate

Bugerman58 committed:
fix upgrade step for migrate - setting analysis and layout

Bugerman58 committed:
add display template for ws; add change num positions

Bugerman58 committed:
fix typo

Bugerman58 commented:
Hi @ramonski and @xispa 

Sorry for a long time of prepare this PR. There were a lot of problems and questions that needed to be solved for the WST to operate under the new conditions.
There were difficulties with the `layout` field (now it's a `template_layout`) for edit and display mode.

Pls check it.

Bugerman58 committed:
update changelog

ramonski reviewed (commented):
I just had a first glance at the code and must say "Wow, what a great piece of work", thanks Anton!
Can you merge the current 2.x branch please, then I will test it with some existing DBs this week.
Thanks again for your work, you are a great support guys!

Bugerman58 committed:
merge 2.x

Bugerman58 commented:
> I just had a first glance at the code and must say "Wow, what a great piece of work", thanks Anton! Can you merge the current 2.x branch please, then I will test it with some existing DBs this week. Thanks again for your work, you are a great support guys!

Thank you for your high appreciation of my work

ramonski committed:
Merge branch '2.x' into migrate-ws-tempaltes-to-dx

ramonski commented:
Hi Anton,
here comes a first feedback of my integration tests:

1. The last upgrade step in `senaite.core` was this one (https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/upgrade/v02_06_000.zcml#L6-L12):

``` xml
  <genericsetup:upgradeStep
      title="SENAITE.CORE 2.6.0: Remove deactivated services from profiles"
      description="Removes deactivated analysis services from Profiles"
      source="2651"
      destination="2652"
      handler=".v02_06_000.remove_inactive_services_from_profiles"
      profile="senaite.core:default"/>
```

Please update make sure you have correctly merged `2.x` and that your upgrade step is `2653`.

---

2. The current implementation adds as many rows as were specified in "Number of Positions":

<img width="1022" alt="Hardness — SENAITE LIMS 2024-10-15 1 PM-56-12" src="https://github.com/user-attachments/assets/84bbeb8f-e5f5-473e-a4c8-21eeb59b028d">

The migrated DX version lacks this functionality (no button available) and rows are added always on the position 1:

<img width="1294" alt="Worksheet Templates — SENAITE LIMS 2024-10-15 2 PM-02-48" src="https://github.com/user-attachments/assets/b0a06cc8-8169-4926-8e52-e4ace28feaea">

---

3. Links to previous created positions do not work:

<img width="1295" alt="Bildschirmfoto 2024-10-15 um 2 05 11 PM" src="https://github.com/user-attachments/assets/3c6ead64-f9af-4f3b-83eb-1b2f74cda04f">

--- 

4. Migrating the demo data Worksheet templates where I added the layout, I get the following traceback:

```
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 288, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module plone.autoform.view, line 40, in __call__
  Module plone.autoform.view, line 62, in _update
  Module z3c.form.group, line 52, in update
  Module z3c.form.group, line 48, in updateWidgets
  Module z3c.form.field, line 277, in update
  Module z3c.form.browser.multi, line 63, in update
  Module z3c.form.browser.widget, line 171, in update
  Module z3c.form.widget, line 496, in update
  Module Products.CMFPlone.patches.z3c_form, line 47, in _wrapped
  Module z3c.form.widget, line 132, in update
  Module z3c.form.widget, line 491, in value
  Module collective.z3cform.datagridfield.datagridfield, line 171, in updateWidgets
  Module z3c.form.widget, line 433, in updateWidgets
  Module z3c.form.widget, line 410, in applyValue
  Module collective.z3cform.datagridfield.datagridfield, line 308, in value
  Module collective.z3cform.datagridfield.datagridfield, line 262, in datagrid_field_set
  Module collective.z3cform.datagridfield.datagridfield, line 311, in updateWidgets
  Module z3c.form.object, line 208, in updateWidgets
  Module collective.z3cform.datagridfield.datagridfield, line 439, in update
  Module plone.z3cform.fieldsets.extensible, line 65, in update
  Module plone.z3cform.patch, line 30, in GroupForm_update
  Module z3c.form.group, line 132, in update
  Module collective.z3cform.datagridfield.datagridfield, line 446, in updateWidgets
  Module z3c.form.form, line 136, in updateWidgets
  Module z3c.form.field, line 277, in update
  Module z3c.form.browser.text, line 36, in update
  Module z3c.form.browser.widget, line 171, in update
  Module Products.CMFPlone.patches.z3c_form, line 47, in _wrapped
  Module z3c.form.widget, line 132, in update
  Module z3c.form.converter, line 116, in toWidgetValue
  Module zope.i18n.format, line 484, in format
TypeError: a float is required
```

## Suggestion

I would remove the "Positions" input field and provide a similar architecture as done for the partition scheme in sample templates:

<img width="1096" alt="Sample Templates — SENAITE LIMS 2024-10-15 2 PM-07-08" src="https://github.com/user-attachments/assets/647ad5a8-6fcd-44ac-9dbf-910d81ede57c">

The partition selector dropdown is also dynamically updated when new partitions are added

<img width="655" alt="Bildschirmfoto 2024-10-15 um 2 09 09 PM" src="https://github.com/user-attachments/assets/6678e4d5-54dc-4164-a7aa-3ea3afe8f940">

ramonski commented:
Minor Issue:

It would be great to find out, why these Analyses Tables are rendered so small on both the Worksheet- and Sample Template settings.

ramonski commented:
Another UI issue:

Clicking on the "Analysis Type" Dropdown completely removes the Worksheet layout:

<img width="1304" alt="Worksheet Templates — SENAITE LIMS 2024-10-15 2 PM-14-01" src="https://github.com/user-attachments/assets/03b37199-6c44-41ef-bc0a-28d889e2f484">

<img width="1296" alt="Worksheet Templates — SENAITE LIMS 2024-10-15 2 PM-14-52" src="https://github.com/user-attachments/assets/15150bdb-51a3-490e-a02b-652f6e10aac4">

ramonski commented:
Maybe related to the UI handler adapter:

When selecting a Reference, the JS Console shows this error:

<img width="1854" alt="Worksheet Templates — SENAITE LIMS 2024-10-15 2 PM-17-12" src="https://github.com/user-attachments/assets/eb3abe92-7baa-4154-a358-25bd39338cef">

ramonski commented:
The "Duplicate of" selector should be only visible or active if the "Analysis Type" was set to "Duplicate":

<img width="1254" alt="My WST — SENAITE LIMS 2024-10-15 2 PM-23-40" src="https://github.com/user-attachments/assets/9e7526c8-8ac0-4695-9fc3-a122c2d07ba6">

Bugerman58 commented:
Hi, Ramon! Thank for you feedback.
> Hi Anton, here comes a first feedback of my integration tests:
> 
> 2. The current implementation adds as many rows as were specified in "Number of Positions":
>
> 3. Links to previous created positions do not work:
> 
_fast answer from me_

very strange interface on your screenshots

maybe you should clear the cache(Shift+Cmd+R)

I'll attach my screenshots a bit later, but this is definitely not how it should be

Bugerman58 commented:
>The last upgrade step in senaite.core was this one (https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/upgrade/v02_06_000.zcml#L6-L12):
  <genericsetup:upgradeStep
      title="SENAITE.CORE 2.6.0: Remove deactivated services from profiles"
      description="Removes deactivated analysis services from Profiles"
      source="2651"
      destination="2652"
      handler=".v02_06_000.remove_inactive_services_from_profiles"
      profile="senaite.core:default"/>
Please update make sure you have correctly merged 2.x and that your upgrade step is 2653.

I'm sure that my version 2653 )

![Screenshot 2024-10-15 at 19 50 00](https://github.com/user-attachments/assets/88c984c7-7750-4ca2-abbf-9df7cc150f68)

Bugerman58 commented:
@ramonski 

this is the correct ui

View mode:

![Screenshot 2024-10-15 at 20 10 34](https://github.com/user-attachments/assets/d524940f-3e7b-4db7-b6ff-03c2ad2ce475)

Edit mode(before change):

![Screenshot 2024-10-15 at 20 10 58](https://github.com/user-attachments/assets/f8cb3aed-076e-4684-9485-5653aa4cb059)


Edit mode(after change num positions):

![Screenshot 2024-10-15 at 20 11 21](https://github.com/user-attachments/assets/c5231e06-fd07-435c-aeb3-43a87efcdbea)

Bugerman58 commented:
> Suggestion
I would remove the "Positions" input field and provide a similar architecture as done for the partition scheme in sample templates:

and as for the SampleTemplates, the `partition id` field does not work correctly when adding and deleting rows:

![Screenshot 2024-10-15 at 20 33 35](https://github.com/user-attachments/assets/c4284e4f-766b-4d5b-a427-164a95b1622f)

ramonski commented:
Hi Anton,
Please apologize, I forgot that additional remotes are not fetched when doing a `git fetch` without the `--all` flag 🙈 
Ok, I'm now on [ac605ef](https://github.com/senaite/senaite.core/pull/2599/commits/ac605ef8383ead39c5f6758a2bf1e1f62bb5a87f) here and installed a fresh SENAITE site.
When importing the demo data, I get the following traceback:
```
2024-10-15 16:42:38,813 ERROR   [senaite.core:67][waitress-0] Traceback (most recent call last):
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/browser/form/adapters/data_import.py", line 62, in handle_data_import
    LoadSetupData(self.context, self.request)()
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/load_setup_data.py", line 119, in __call__
    adapter(self, workbook, self.dataset_project, self.dataset_name)
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 144, in __call__
    self.Import()
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 2005, in Import
    folder = self.context.bika_setup.bika_worksheettemplates
AttributeError: 'RequestContainer' object has no attribute 'bika_worksheettemplates'
```

Bugerman58 commented:
> Hi Anton, Please apologize, I forgot that additional remotes are not fetched when doing a `git fetch` without the `--all` flag 🙈 Ok, I'm now on [ac605ef](https://github.com/senaite/senaite.core/pull/2599/commits/ac605ef8383ead39c5f6758a2bf1e1f62bb5a87f) here and installed a fresh SENAITE site. When importing the demo data, I get the following traceback:
> 
> ```
> 2024-10-15 16:42:38,813 ERROR   [senaite.core:67][waitress-0] Traceback (most recent call last):
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/browser/form/adapters/data_import.py", line 62, in handle_data_import
>     LoadSetupData(self.context, self.request)()
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/load_setup_data.py", line 119, in __call__
>     adapter(self, workbook, self.dataset_project, self.dataset_name)
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 144, in __call__
>     self.Import()
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 2005, in Import
>     folder = self.context.bika_setup.bika_worksheettemplates
> AttributeError: 'RequestContainer' object has no attribute 'bika_worksheettemplates'
> ```

Excellent. You managed to achieve the correct UI behavior?

`exportimport` and `demodata` is not so important now and can be adjusted

ramonski commented:
> > Suggestion
> > I would remove the "Positions" input field and provide a similar architecture as done for the partition scheme in sample templates:
> 
> and as for the SampleTemplates, the `partition id` field does not work correctly when adding and deleting rows:
> 
> ![Screenshot 2024-10-15 at 20 33 35](https://private-user-images.githubusercontent.com/15684023/376633558-c4284e4f-766b-4d5b-a427-164a95b1622f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg5OTk3ODMsIm5iZiI6MTcyODk5OTQ4MywicGF0aCI6Ii8xNTY4NDAyMy8zNzY2MzM1NTgtYzQyODRlNGYtNzY2Yi00ZDViLWE0MjctMTY0YTk1YjE2MjJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE1VDEzMzgwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlY2ZjYzcyYjQ4YzlhMmY5YjRiNjQyNmZhNDcyMjYyY2MxODRlZWFkOGU2OWQ1MTMwZWQzMTY4Y2M1NzZkNDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.TBJ2p3HrcdanGYFDFnv97yXIkXoUsZYNp1bo5CVDl8w)

Yes, you are right. I was able to reproduce the bug and this should be definitely fixed.

Bugerman58 commented:
> When importing the demo data, I get the following traceback:

I'll try to fix it this week

ramonski commented:
> > When importing the demo data, I get the following traceback:
> 
> I'll try to fix it this week

Great, thanks Anton! This makes testing a bit easier.

The UI looks great now and works like a charm!

My test setup has 2 instances, one with the current `2.x` branch and one with your branch:

**Old UI**

<img width="1832" alt="Hardness — SENAITE LIMS 2024-10-16 7 AM-53-28" src="https://github.com/user-attachments/assets/07509e6d-02e3-4a6f-b9bf-d05bc02230a7">

**New UI**

<img width="1270" alt="Hardness — SENAITE LIMS 2024-10-16 7 AM-49-48" src="https://github.com/user-attachments/assets/d537ecfe-05b6-4330-8eb5-2de12a15e157">

During my tests, I noted a side-effect **on both versions** when I try to use these Templates:

**Current 2.x Branch**

<img width="1831" alt="WS-001 — SENAITE LIMS 2024-10-16 7 AM-55-26" src="https://github.com/user-attachments/assets/93d01bb1-4264-436f-9fcd-7859b1f3916e">

**Your feature Branch**

<img width="1830" alt="WS-001 — SENAITE LIMS 2024-10-16 7 AM-56-48" src="https://github.com/user-attachments/assets/a82fb706-4986-4050-be34-69dc5cddc0c7">

While it is good that it actually behaved the same, I asked myself why the Blank and Reference Sample were not assigned, why Position 3 is missing and Position 4 is a Reassignable slot.

In both versions there are only Reference Definitions present, which are those listed in the WST dropdowns:

<img width="997" alt="Reference Definitions — SENAITE LIMS 2024-10-16 8 AM-07-49" src="https://github.com/user-attachments/assets/82a86db6-6b52-45a6-8cf4-21741d38ed4a">

But there are no Reference Samples below Suppliers present:

<img width="1829" alt="Reference Samples cc — SENAITE LIMS 2024-10-16 8 AM-08-59" src="https://github.com/user-attachments/assets/fb7804f8-c6ab-44f4-b45f-6d86952d0905">

@xispa: This is something we probably should address when touching Worksheets the next time.

Anyhow, just a reminder at this place for now.

ramonski commented:
> Minor Issue:
> 
> It would be great to find out, why these Analyses Tables are rendered so small on both the Worksheet- and Sample Template settings.

This PR fixes the rendering issue for listing widgets: https://github.com/senaite/senaite.core/pull/2625

ramonski commented:
Btw, just because I've seen it now again.
With https://github.com/senaite/senaite.core/pull/2588 you should be able to style the "Set" button like in the old version. But it is perfectly fine how it looks at the moment as well.

ramonski commented on src/senaite/core/browser/form/adapters/worksheettemplate.py:
```diff
@@ -18,37 +18,213 @@
+                mapping={"dup": dup_pos})
+            )
+            self.add_error_field(FIELD_TYPE.format(index), msg)
+            return False
```
Good check!

<img width="937" alt="ABX Pentra 60 — SENAITE LIMS 2024-10-16 2 PM-16-58" src="https://github.com/user-attachments/assets/91f39d70-5f8a-4c4f-a207-27a289d2f7b7">

Bugerman58 replied:
Thanks! Idea and text  copy-paste from old code  🤫

ramonski commented on src/senaite/core/content/worksheettemplate.py:
```diff
@@ -0,0 +1,570 @@
+
+def default_layout_positions(count_positions=None, start_pos=0):
+    """
+    """
```
Please add a docstring here

ramonski commented on src/senaite/core/content/worksheettemplate.py:
```diff
@@ -0,0 +1,570 @@
+    elif request:
+        params = dict(parse_qs(request["QUERY_STRING"]))
+        values = params.get("num_positions")
+        num_positions = int(values[0]) if values else 0
```
Add a comment here, that you are called in this case from the add form adapter (`senaite.core.browser.form.wstemplate`) with  `++add++WorksheetTemplate?num_positions={}`

ramonski commented on src/senaite/core/content/worksheettemplate.py:
```diff
@@ -0,0 +1,570 @@
+    request = api.get_request()
+    if request:
+        params = dict(parse_qs(request["QUERY_STRING"]))
+        values = params.get("num_positions")
```
duplicate code here. Please add a function e.g. `get_query_string_by_name(name)` into a suitable place.

Bugerman58 replied:
oh, sure.

xispa commented:
> @xispa: This is something we probably should address when touching Worksheets the next time.

I was thinking that maybe it would only make sense to allow the selection of a sample template with References when there are counterpart QC samples registered and valid. However, the QC Samples hold reference values that apply to tests and we cannot foresee that the resulting worksheet will contain analyses for all tests defined in the template... Therefore, I think best would be to simply place an alert panel stating something like this:


> [!WARNING]  
> *This slot is intended to contain a quality control sample for the 'Trace Metals 9' reference definition; however, no active or valid quality control samples were identified.*


We could also place a reference widget with [Add] button inside that very same sample to easily add such a quality control sample when available.

This is for another PR.

ramonski commented on src/senaite/core/content/worksheettemplate.py:
```diff
@@ -0,0 +1,570 @@
+            u"description_worksheettemplate_multiple_user_of_instrument",
+            default=u"If unchecked, Lab Managers won't be able to assign "
+                    u"the same Instrument more than one Analyses while "
+                    u"creating a Worksheet."
```
I'm confused by the current text we have here.
@xispa: Can you remember what the purpose of this checkbox was?

ramonski replied:
Ah ok, found it here:
https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/browser/worksheet/templates/results.pt#L255-L258
And I guess it has absolutely no function, or at least I cannot find a JS that queries the ID `instrument_multiple_use`

Shall we simply remove this field?

xispa replied:
Pheew, this comes from #133 ... really old code... and not sure what was the expected functionality... Yes, remove the field, please

Bugerman58 replied:
I agree. Done.

ramonski commented on src/senaite/core/content/worksheettemplate.py:
```diff
@@ -0,0 +1,570 @@
+            self.setNumOfPositions(num_pos)
+
+    # BBB: AT schema field property
+    Layout = property(getTemplateLayout, setTemplateLayout)
```
Please add the method `getLayout` with a deprecation decorator for backwards compatibility here

Bugerman58 replied:
Hi Ramon!
Can not declare method `getLayout` because this method invoke from parent Plone view and throw exception:
```shell
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 264, in publish
  Module ZPublisher.BaseRequest, line 450, in traverse
  Module plone.rest.patches, line 15, in __before_publishing_traverse__
  Module Products.CMFCore.DynamicType, line 164, in __before_publishing_traverse__
  Module Products.CMFDynamicViewFTI.fti, line 239, in queryMethodID
  Module Products.CMFDynamicViewFTI.fti, line 201, in defaultView
  Module Products.CMFPlone.PloneTool, line 696, in browserDefault
  Module bika.lims, line 195, in wrapper
TypeError: getLayout() got an unexpected keyword argument 'check_exists'
```

[`Products.CMFPlone.PloneTool:695`](https://github.com/plone/Products.CMFPlone/blob/master/Products/CMFPlone/PloneTool.py#L695)

`bika.lims, line 195` is deprecation decorator

_this example from my instance_

Bugerman58 replied:
That is why the `layout` field was renamed to `template_layout`: https://github.com/senaite/senaite.core/pull/2599#issuecomment-2376098725

Bugerman58 replied:
without annotation of deprecation:
```shell
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 264, in publish
  Module ZPublisher.BaseRequest, line 450, in traverse
  Module plone.rest.patches, line 15, in __before_publishing_traverse__
  Module Products.CMFCore.DynamicType, line 164, in __before_publishing_traverse__
  Module Products.CMFDynamicViewFTI.fti, line 239, in queryMethodID
  Module Products.CMFDynamicViewFTI.fti, line 201, in defaultView
  Module Products.CMFPlone.PloneTool, line 696, in browserDefault
TypeError: getLayout() got an unexpected keyword argument 'check_exists'
```
...the same

Bugerman58 replied:
@ramonski pls, check that is above

Bugerman58 replied:
...

I think that no possibility use this method

ramonski commented on src/senaite/core/content/worksheettemplate.py:
```diff
@@ -0,0 +1,570 @@
+        mutator(self, records)
+
+    # BBB: AT schema field property
+    Services = property(getServices, setServices)
```
Please add the method `getRawService` with a deprecation decorator for backwards compatibility here

ramonski commented on src/senaite/core/z3cform/widgets/datagrid/datagrid.py:
```diff
@@ -66,6 +67,11 @@ def createObjectWidget(self, idx):
+        if self.mode in self.templates.keys():
+            return ViewPageTemplateFile(self.templates[self.mode])(self)
+        return super(DataGridWidget, self).render()
+
```
cool!

Bugerman58 replied:
Tnx! )

ramonski commented on src/senaite/core/z3cform/widgets/datagrid/ws_template_datagrid_display.pt:
```diff
@@ -0,0 +1,63 @@
+              <?python
+                blank_ref.mode = "display"
+                control_ref.mode = "display"
+              ?>
```
Oh wow, I haven't seen/used this construct before!

Bugerman58 replied:
nano-hack for template of widget 😉

ramonski reviewed (changes requested):
Great work Anton and I see the efforts you have put into this PR!
I added some comments (also for @xispa) and I guess when we have the updated setupdata, then we can merge it into Core!

ramonski committed:
Merge branch '2.x' into migrate-ws-tempaltes-to-dx

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-tempaltes-to-dx

Bugerman58 committed:
delete unused field EnableMultipleUseOfInstrument

Bugerman58 committed:
refactoring for duplicate code

Bugerman58 committed:
fix exportimport setupdata for WST

Bugerman58 committed:
update demodata

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-tempaltes-to-dx

Bugerman58 committed:
add old methods for backward compatibility with annotation of deprecated

Bugerman58 committed:
delete conflicting method

ramonski commented:
Thanks for the changes Anton!
I tested the migration of an existing Database and everything seems to work without issues.
Unfortunately, the demo data import still raises an error:

```
2024-10-21 20:21:51,307 ERROR   [senaite.core:67][waitress-1] Traceback (most recent call last):
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/browser/form/adapters/data_import.py", line 62, in handle_data_import
    LoadSetupData(self.context, self.request)()
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/load_setup_data.py", line 119, in __call__
    adapter(self, workbook, self.dataset_project, self.dataset_name)
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 144, in __call__
    self.Import()
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 2033, in Import
    obj.setTemplateLayout(self.wst_layouts[title])
AttributeError: Worksheet_Templates instance has no attribute 'wst_layouts'
```

Bugerman58 committed:
fix import worksheet templates

Bugerman58 commented:
> Thanks for the changes Anton! I tested the migration of an existing Database and everything seems to work without issues. Unfortunately, the demo data import still raises an error:
> 
> ```
> 2024-10-21 20:21:51,307 ERROR   [senaite.core:67][waitress-1] Traceback (most recent call last):
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/browser/form/adapters/data_import.py", line 62, in handle_data_import
>     LoadSetupData(self.context, self.request)()
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/load_setup_data.py", line 119, in __call__
>     adapter(self, workbook, self.dataset_project, self.dataset_name)
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 144, in __call__
>     self.Import()
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 2033, in Import
>     obj.setTemplateLayout(self.wst_layouts[title])
> AttributeError: Worksheet_Templates instance has no attribute 'wst_layouts'
> ```

oh, sorry

now checked on my instance and WST import

ramonski commented:
I get now this error on my side:
```
2024-10-23 08:45:41,868 ERROR   [senaite.core:67][waitress-0] Traceback (most recent call last):
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/browser/form/adapters/data_import.py", line 62, in handle_data_import
    LoadSetupData(self.context, self.request)()
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/load_setup_data.py", line 113, in __call__
    in list(getAdapters((self.context, ), ISetupDataImporter))]
  File "/Users/rbartl/develop/buildout/eggs/zope.interface-5.5.2-py2.7-macosx-15.0-x86_64.egg/zope/interface/registry.py", line 377, in getAdapters
    adapter = factory(*objects)
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 1969, in __init__
    super(Worksheet_Templates, self).__init__(context)
TypeError: super() argument 1 must be type, not classobj
```

ramonski commented:
I guess it is because the base class does not inherit from `object`:
```

class WorksheetImporter:

    """Use this as a base, for normal tabular data sheet imports.
    """

    def __init__(self, context):
        self.adapter_context = context

```

ramonski commented:
Ok, that seems to work. Please change the base class to this:
```python
class WorksheetImporter(object):
    """Use this as a base, for normal tabular data sheet imports.
    """
```

I'm happy when we have this Worksheet importer abandoned...

ramonski commented:
The demo data worked now, but the imported worksheet template seems to be misconfigured:

![image](https://github.com/user-attachments/assets/3bddc122-ea82-4883-863b-b49bae1a9cf0)

Bugerman58 commented:
> I get now this error on my side:
>  super(Worksheet_Templates, self).__init__(context)
> TypeError: super() argument 1 must be type, not classobj

oh, yes. this code was added after checked import. 

> The demo data worked now, but the imported worksheet template seems to be misconfigured:

yes, I understand why this is so - the import file uses values ​​from the dictionary, not keys. I will add a check for this.

Bugerman58 committed:
fix after review

Bugerman58 commented:
> The demo data worked now, but the imported worksheet template seems to be misconfigured:
> 

Hi Ramon! Sorry for my inadvertence. I hope now that will working correctly. =|

ramonski commented:
Hi Anton, Thanks for the fixtures!
I guess we're almost there and I only found the following smaller issues:

1. After demo data installation, the fields show "Constraint not satisfied". But the messages disappear after saving:

after demo-data import:

<img width="946" alt="Hardness — SENAITE LIMS 2024-10-26 8 PM-29-29" src="https://github.com/user-attachments/assets/9404fdbc-351b-4b6e-8e9e-bcbfa448a9ee">

after save:

<img width="943" alt="Hardness — SENAITE LIMS 2024-10-26 8 PM-30-27" src="https://github.com/user-attachments/assets/b558fee9-db9b-4059-b861-4b78748db120">


2. Removing the slot number and pressing the "Set" button raises this error:

<img width="864" alt="Hardness — SENAITE LIMS 2024-10-26 8 PM-31-19" src="https://github.com/user-attachments/assets/56a3489c-3626-4346-aac4-f4a8e86c5e71">

3. Something strange but worth to mention here as well (but probably the same thing as already mentioned above) – When I create 4 demo Samples with all the "Trace Metals" profile applied and create afterwards a WS with the Template from the demo-data, I get this result:

<img width="1688" alt="WS-001 — SENAITE LIMS 2024-10-26 8 PM-36-23" src="https://github.com/user-attachments/assets/f3ec291d-81e8-4a21-ba31-44a858b7f5f1">

4. Creating a new Worksheet Template (or edit an existing) gives me this error in the JS console:

<img width="1714" src="https://github.com/user-attachments/assets/52dff107-0f42-4251-bed6-5a1d6fa0ce30">


To summarize:

In my opinion, the traceback and the JS error should be fixed. The rest of the issues seem to be minor or belong to another PR. Let me know what you think (also @xispa).

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-tempaltes-to-dx

Bugerman58 committed:
fix update num_position if it is empty

Bugerman58 committed:
fix import wst with definitions title

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-tempaltes-to-dx

Bugerman58 commented:
> 1. After demo data installation, the fields show "Constraint not satisfied". But the messages disappear after saving:

Solved. Added check for `blank_ref` and `control_ref` fields on value of title of definitions.

But this message(`Constraint not satisfied`) may appear for the `dup` field - this now in work.

> 2. Removing the slot number and pressing the "Set" button raises this error:

Fixed.

xispa commented:
> In my opinion, the traceback and the JS error should be fixed. The rest of the issues seem to be minor or belong to another PR. Let me know what you think (also @xispa).

I Will do functional testing today. Worth to mention that I played around worksheet templates yesterday with a heavy-customized instance and I stumbled upon some issues. Will take those issues into account when playing around.

xispa commented:
Except the "Constraint not satisfied" thing, the rest seems to work properly to me.

xispa commented:
Just wanted to say good work @Bugerman58 , you rock!

Bugerman58 commented:
> Just wanted to say good work @Bugerman58 , you rock!

Thanks! 😊

ramonski committed:
Merge branch '2.x' into migrate-ws-tempaltes-to-dx

ramonski reviewed (approved):
Thanks Anton! Besides the JS error, everything work perfectly.
I'm going to add another PR and try to fix it in the `editform.js`.
One step further towards Python 3;)
